libs/ledger/src/time_based_transaction_status_cache.cpp

```diff
@@ -0,0 +1,160 @@
+
+#include "network/generics/milli_timer.hpp"
+#include "time_based_transaction_status_cache.hpp"
+
+namespace fetch {
+namespace ledger {
+namespace {
+
+constexpr std::chrono::hours   LIFETIME{24};
+constexpr std::chrono::minutes INTERVAL{5};
+
+using TxStatus = TimeBasedTransactionStatusCache::TxStatus;
+
+}  // namespace
+
+/**
+ * Query the status of a specified transaction
+ *
+ * @param digest The digest of the transaction
+ * @return The status object associated for this transaction
+ */
+TxStatus TimeBasedTransactionStatusCache::Query(Digest digest) const
+{
+  {
+    FETCH_LOCK(mtx_);
+
+    auto const it = cache_.find(digest);
+    if (cache_.end() != it)
+    {
+      return it->second.status;
+    }
+  }
+
+  return {};
+}
+
+/**
+ * Update the status of a transaction with the specified status enum
+ *
+ * @param digest The transaction to be updated
+ * @param status The status value that should be set
+ */
+void TimeBasedTransactionStatusCache::Update(Digest digest, TransactionStatus status)
+{
+  auto const now{clock_->Now()};
+
+  FETCH_LOCK(mtx_);
+
+  if (TransactionStatus::EXECUTED == status)
+  {
+    FETCH_LOG_WARN("TransactionStatusCache",
+                   "Using inappropriate method to update contract "
+                   "execution result. (tx digest: 0x",
+                   digest.ToHex(), ")");
+
+    throw std::runtime_error(
+        "TransactionStatusCache::Update(...): Using inappropriate method to update"
+        "contract execution result");
+  }
+
+  auto it = cache_.find(digest);
+
+  if (it == cache_.end())
+  {
+    cache_.emplace(digest, CacheEntry{TxStatus{status}, now});
+  }
+  else
+  {
+    it->second.status.status = status;
+  }
+
+  PruneCacheIfNecessary(now);
+}
+
+/**
+ * Update the contract execution result for the specified transaction
+ *
+ * @param digest The transaction to be updated
+ * @param exec_result The contract execution result
+ */
+void TimeBasedTransactionStatusCache::Update(Digest digest, ContractExecutionResult exec_result)
+{
+  static constexpr TransactionStatus EXECUTED_STATUS{TransactionStatus::EXECUTED};
+
+  auto const now{clock_->Now()};
+
+  FETCH_LOCK(mtx_);
+
+  // update the cache
+  auto it = cache_.find(digest);
+  if (it == cache_.end())
+  {
+    FETCH_LOG_DEBUG("TransactionStatusCache",
+                    "Updating contract execution status for transaction "
+                    "which is missing in the tx status cache. (tx digest: 0x",
+                    digest.ToHex(), ")");
+
+    cache_.emplace(digest, CacheEntry{TxStatus{EXECUTED_STATUS, exec_result}, now});
+  }
+  else
+  {
+    it->second.status.status               = EXECUTED_STATUS;
+    it->second.status.contract_exec_result = exec_result;
+  }
+
+  PruneCacheIfNecessary(now);
+}
+
+void TimeBasedTransactionStatusCache::PruneCache(Timestamp const &until)
+{
+  fetch::generics::MilliTimer timer{"TxStatusCache::Prune"};
```

Comment:
Update timer name.

Reply:
I think that the name is good here to be honest.

Reply:
But we now have two different status caches - as it stands, the logs will not be explicit whether `TxStatusCache` refers to the time-based one or the persistent one.

Reply:
The persistent one doesn't have a prune (any will never have one I would guess). This is why this seemed like a fair name